Add initial main.go code with main implementation

tremaineeto opened the pull request:
- Utilizes in-cluster-configuration (clusterrole and clusterrolebinding)
  to connect to Kubernetes resources within the same cluster

- Goes through each pod and checks re-start count, cascade deleting any
  deployment that has a pod with 100 or more restarts

tremaineeto committed:
Add initial main.go code with main implementation

- Utilizes in-cluster-configuration (clusterrole and clusterrolebinding)
  to connect to Kubernetes resources within the same cluster

- Goes through each pod and checks re-start count, cascade deleting any
  deployment that has a pod with 100 or more restarts

tremaineeto commented:
This is just to get the ball rolling. There are a lot of improvements to be made to this, not limited to:

- Limiting the if statement nesting
- Refining variables/objects being used
- Code commenting
- Better error handling
- More descriptive logging (and using something other than println, probably)
- Externalized configuration

Any contributions, examples, or suggestions to achieve these would be greatly appreciated!

rymonroe commented on main.go:
```diff
@@ -0,0 +1,74 @@
+
+		for i := 0; i < len(pods.Items); i++ {
+			for j := 0; j < len(pods.Items[i].Status.ContainerStatuses); j++ {
+				if pods.Items[i].Status.ContainerStatuses[j].State.Waiting != nil {
```
It may be more readable to take the checks from lines 38-41 and extract them into a method. Just a suggestion.

timjonesdev committed:
clean up go.mod

tremaineeto commented on main.go:
```diff
@@ -25,34 +38,36 @@ func main() {
 				if waiting != nil {
 					reason := waiting.Reason
-					if reason == "CrashLoopBackOff" && status.RestartCount > restartThreshold {
+					if stringInSlice(reason, waitingReasons) && status.RestartCount > restartThreshold {
```
@jonathantronson Here is an idea I had for choosing if the waiting reason is in a slice of possibilities. What do you think?

timjonesdev replied:
https://stackoverflow.com/a/10486196/6520477

I like this idea. I went looking for an "includes" or "contains" type of method, and came up with that SO question. It's probably more efficient to do it the way they discuss in the SO question. Since this loop touches every pod, we might like that efficiency.

tremaineeto commented:
Updated with @timjonesdev's branch and with suggestions. Any new feedback welcome.

timjonesdev commented on main.go:
```diff
@@ -25,34 +38,36 @@ func main() {
 					reason := waiting.Reason
-					if reason == "CrashLoopBackOff" && status.RestartCount > restartThreshold {
+					if stringInSlice(reason, waitingReasons) && status.RestartCount > restartThreshold {
 						fmt.Println(item.Namespace + "/" + item.Name + " has " +
```
I would convert this to Printf also.

tremaineeto replied:
Is this because it is easier to maintain over time? Genuinely curious if Printf() is better than Println() generally. I didn't see anything online about how (in Go specifically) if it is more efficient. I don't mind switching it over, but wondering what I should default to in the future.

timjonesdev commented:
Overall, I like the changes so far.

jonathantronson commented:
Look at using labels for your continue statements in your nested for loops

Commit:
Add labels to continue statements

tremaineeto committed:
Changed a Println to a Printf

Commit:
Add waiting reasons map

- Presumably better for complexity

tremaineeto commented on main.go:
```diff
@@ -15,19 +15,15 @@ func (b bin) String() string {
 	NUMBER_RESTARTS := 100
 
-	waitingReasons := []string{"CrashLoopBackOff", "ErrImagePull", "Completed", "Failed"}
+	waitingReasons := map[string]struct{}{
```
@timjonesdev Here I'm trying out the map and then below the check for the key within the map. Does this look about right?

tremaineeto commented on main.go:
```diff
@@ -49,15 +45,20 @@ func main() {
-						fmt.Printf("%s / %s has %s restarts, which is over the %s restart limit.", item.Namespace,
-							item.Name, strconv.Itoa(int(status.RestartCount)), strconv.Itoa(int(restartThreshold)))
+					_, reasonInWaitingReasons := waitingReasons[reason]
+					if reasonInWaitingReasons || (reason == "CrashLoopBackOff" && status.RestartCount > restartThreshold) {
```
I want to potentially brainstorm this if check. It makes sense to me to have "CrashLoopBackOff" in the map of reasons that are bad, but in that case I want to concisely capture these requirements for deletion:

- If the waiting reason is in the map and is anything but CrashLoopBackOff, then proceed with trying to delete
OR
- If the waiting reason is in the map AND is CrashLoopBackOff, then check if the number of restarts is over the threshold. If it is, then proceed with trying to delete

I can easily make the if checks for these, but do you think this current way of keeping CrashLoopBackOff separate from the map is better? Or, rather, is there some better way to organize these? Any thoughts?

tremaineeto committed:
Add shields to README

tremaineeto committed:
Merge pull request #3 from att-cloudnative-labs/markdown

Add shields to README

tremaineeto committed:
Merge branch 'master' into initial-code

timjonesdev committed:
add Dockerfile
Add note about running in-cluster only

timjonesdev committed:
implement an example of the interface discussion I had with Tremaine

timjonesdev committed:
implement a different example of the interface discussion I had with Tremaine

bhaskar-reddy commented on main.go:
```diff
@@ -0,0 +1,97 @@
+		"ImagePullBackOff": {},
+	}
+
+	config, err := rest.InClusterConfig()
```
I forgot where I saw this, but combining method call and error check into one, would make it more succinct.. 
For ex., lines 28-31 can be written as:	
```
if config, err := rest.InClusterConfig(); err != nil {
	panic(err.Error())
}
```
Same for all places where we are doing the immediate err != nil check.

tremaineeto replied:
Thanks for showing me this way of writing it. I've changed it in my local.

timjonesdev committed:
Add Viper for external config
Create structs to hold configs
Set up system for assigning functions based on configs

timjonesdev committed:
Refactor and fill out delete functions

timjonesdev committed:
Add new files to the Dockerfile

timjonesdev committed:
refactor Dockerfile

jonathantronson reviewed (commented):
Looks good to me.

timjonesdev committed:
Merge pull request #4 from att-cloudnative-labs/feature/dockerfile

add Dockerfile

Commit:
Add comments

- Add logging statements
- Consolidate one if check

Commit:
Update print statement grammar

gbraxton commented on go.sum:
```diff
@@ -0,0 +1,67 @@
+cloud.google.com/go v0.34.0/go.mod h1:aQUYkXzVsufM+DwF1aE+0xfcU+56JwCaLick0ClmMTw=
```
Should this file be ignored?

timjonesdev replied:
I believe it should be included, and not ignored. https://www.reddit.com/r/golang/comments/97dxj1/should_i_add_the_gosum_file_to_the_repo_of_a/

gbraxton reviewed (approved):
Looks Good

timjonesdev committed:
Merge pull request #6 from att-cloudnative-labs/feature/configurable_function_map

Feature/configurable function map
